pkg/volume/csi/csi_client.go

```diff
@@ -841,3 +854,109 @@ func (c *csiClientGetter) Get() (csiClient, error) {
 	c.csiClient = csi
 	return c.csiClient, nil
 }
+
+func (c *csiDriverClient) NodeSupportsVolumeStats(ctx context.Context) (bool, error) {
+	klog.V(4).Info(log("calling NodeGetCapabilities rpc to determine if NodeSupportsVolumeStats"))
+
+	if c.nodeV1ClientCreator != nil {
+		return c.nodeSupportsVolumeStatsV1(ctx)
+
+	}
+	return false, fmt.Errorf("failed to call NodeSupportsVolumeStats. nodeV1ClientCreator is nil")
+}
+
+func (c *csiDriverClient) nodeSupportsVolumeStatsV1(ctx context.Context) (bool, error) {
+	nodeClient, closer, err := c.nodeV1ClientCreator(c.addr)
+	if err != nil {
+		return false, err
+	}
+	defer closer.Close()
+
+	req := &csipbv1.NodeGetCapabilitiesRequest{}
+	resp, err := nodeClient.NodeGetCapabilities(ctx, req)
+	if err != nil {
+		return false, err
+	}
+
+	capabilities := resp.GetCapabilities()
+
+	if capabilities == nil {
+		return false, nil
+	}
+	for _, capability := range capabilities {
+		if capability.GetRpc().GetType() == csipbv1.NodeServiceCapability_RPC_GET_VOLUME_STATS {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (c *csiDriverClient) NodeGetVolumeStats(ctx context.Context, volID string, targetPath string) (map[string]usageCount, error) {
+	klog.V(4).Info(log("calling NodeGetVolumeStats rpc: [volid=%s, target_path=%s", volID, targetPath))
+	if volID == "" {
+		return nil, errors.New("missing volume id")
+	}
+	if targetPath == "" {
+		return nil, errors.New("missing target path")
+	}
+
+	if c.nodeV1ClientCreator != nil {
+		return c.nodeGetVolumeStatsV1(ctx, volID, targetPath)
+	}
+
+	return nil, fmt.Errorf("failed to call NodeGetVolumeStats. nodeV1ClientCreator is nil")
+}
+
+func (c *csiDriverClient) nodeGetVolumeStatsV1(
+	ctx context.Context,
+	volID string,
+	targetPath string,
+) (map[string]usageCount, error) {
+
+	var unit csipbv1.VolumeUsage_Unit
+	nodeClient, closer, err := c.nodeV1ClientCreator(c.addr)
+	if err != nil {
+		return nil, err
+	}
+	defer closer.Close()
+
+	req := &csipbv1.NodeGetVolumeStatsRequest{
+		VolumeId:   volID,
+		VolumePath: targetPath,
+	}
+
+	resp, err := nodeClient.NodeGetVolumeStats(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	usages := resp.GetUsage()
+
+	usageMap := make(map[string]usageCount, 2)
+	if usages == nil {
+		return nil, nil
+	}
+	for _, usage := range usages {
+
+		unit = usage.GetUnit()
+
+		switch unit.String() {
+		case "BYTES":
+			var ucb usageCount
+			ucb.available = usage.GetAvailable()
+			ucb.total = usage.GetTotal()
+			ucb.used = usage.GetUsed()
+			usageMap["BYTES"] = ucb
+		case "INODES":
+			var uci usageCount
+			uci.available = usage.GetAvailable()
+			uci.total = usage.GetTotal()
+			uci.used = usage.GetUsed()
+			usageMap["INODES"] = uci
+		case "default":
```

Comment:
I meant `default:` case, not literally "default" string. It is unlikely we will get "default" string there.

Reply:
ah.. zombie eyes , corrected it.